File: 203. 找相同/203.py

def work1(n, s):
    s = list(map(int, s))
    counter = 1
    ans = s[0]
    for i in range(1, len(s)):
        now = s[i]
        if now == ans:
            counter += 1
        else:
            counter -= 1
        if counter <= 0:
            counter = 1
            ans = now
    return ans


def work2(n, s):
    s = list(map(int, s))
    d = dict()
    for x in s:
        if x in d:
            d[x] += 1
        else:
            d[x] = 1
        if d[x] * 2 > n:
            return x


# -- author: lijw --
if __name__ == '__main__':
    while True:
        try:
            n = int(input())
            s = input().split(' ')
        except EOFError:
            break
        print(work2(n, s))

File: 203. 找相同/203.cpp
#include<cstdio>
#include<cstdlib>
#include<cstring>
#include<algorithm>

using namespace std;

#define f(i,a,b) for(i=(a);i<=(b);i++)
#define fd(i,a,b) for(i=(a);i>=(b);i--)
#define fff(i,a,b,k) for(i=(a);i<=(b);i+=k)
#define fffd(i,a,b,k) for(i=(a);i>=(b);i-=k)
#define NN printf("\n")
#define TT printf("\t")
#define SP printf(" ")

#define LEN 1000010
#define INF 0x7fffffff

// -- author: lijw --
int main()
{
    int n, i, counter, ans,now;
    while(scanf("%d",&n)!=EOF){
        counter=1;
        scanf("%d", &ans);
        f(i,2,n){
            scanf("%d", &now);
            if(now==ans)counter++;
            else counter--;
            if(counter<=0){
                counter=1;
                ans=now;
            }
        }
        printf("%d\n", ans);
    }
    return 0;
}

